src/main/java/org/mariadb/jdbc/internal/util/ByteBuffer.java

```diff
@@ -0,0 +1,271 @@
+				this.pos += len;
+			}
+		}
+	}
+	
+	public void put(byte[] bytes) {
+		if (bytes == null || bytes.length ==0) {
+			return;
+		}
+		ByteBufferArray array = this.buffer;
+		if (array.remaining() > bytes.length) {
+			System.arraycopy(bytes, 0, array.buf, array.pos, bytes.length);
+			array.pos += bytes.length;
+			this.pos += bytes.length;
+		} else {
+			this.buffer = new ByteBufferArray(bytes);
+			this.buffers.add(this.buffer);
+			//allocate(DEFAULT_PAGE_SIZE);
+			this.pos += bytes.length;
+		}
+	}
+
+	public void putShort(short x) {
+		buffer.buf[buffer.pos++] = (byte) x;
+		buffer.buf[buffer.pos++] = (byte) (x >> 8);
+		this.pos += 2;
+	}
+
+	public void putInt(int x) {
+		if (this.buffer.pos + 4 > this.buffer.buf.length) {
+			allocate(DEFAULT_PAGE_SIZE);
+		}
+		ByteBufferArray buffer = this.buffer;
+		buffer.buf[buffer.pos++] = (byte) x;
+		buffer.buf[buffer.pos++] = (byte) (x >> 8);
+		buffer.buf[buffer.pos++] = (byte) (x >> 16);
+		buffer.buf[buffer.pos++] = (byte) (x >> 24);
+		this.pos += 4;
+	}
+
+
+	public void putLong(long x) {
+		if (this.buffer.pos + 4 > this.buffer.buf.length) {
+			allocate(DEFAULT_PAGE_SIZE);
+		}
+		ByteBufferArray buffer = this.buffer;
+		buffer.buf[buffer.pos++] = (byte) x;
+		buffer.buf[buffer.pos++] = (byte) (x >> 8);
+		buffer.buf[buffer.pos++] = (byte) (x >> 16);
+		buffer.buf[buffer.pos++] = (byte) (x >> 24);
+		buffer.buf[buffer.pos++] = (byte) (x >> 32);
+		buffer.buf[buffer.pos++] = (byte) (x >> 40);
+		buffer.buf[buffer.pos++] = (byte) (x >> 48);
+		buffer.buf[buffer.pos++] = (byte) (x >> 56);
+		this.pos += 8;
+	}
+
+	public void put(byte x) {
+		if (this.buffer.pos + 1 > this.buffer.buf.length) {
+			allocate(DEFAULT_PAGE_SIZE);
+		}
+		this.buffer.buf[this.buffer.pos++] =  x;	
+		this.pos += 1;
+	}
+	
+	public void putString(String str) {
+		Utf8.write(this, UnsafeString.getChars(str), 0 , str.length());
+	}
+
+	public void putBytes(byte theByte, int count) {
+		if (this.buffer.pos + count > this.buffer.buf.length) {
+			allocate(DEFAULT_PAGE_SIZE);
+		}
+		ByteBufferArray buffer = this.buffer;
+		for (int i = 0; i < count; i++) {
+			buffer.buf[buffer.pos++] = theByte;
+        }
+		this.pos += count;
+	}
+	
+	public void putStream(InputStream is, long readLength) throws IOException {
+		ByteBufferArray array = this.buffer;
+		if (array.pos + readLength < array.buf.length) {
+			is.read(array.buf, array.pos, (int)readLength);
+			this.pos += (int)readLength;
+		} else {
+			long remainingReadLength = readLength;
+			int read;
+			while (remainingReadLength > 0) {
+				allocate(DEFAULT_PAGE_SIZE_STREAM); 
+				read = is.read(this.buffer.buf, 0, Math.min((int)remainingReadLength, DEFAULT_PAGE_SIZE_STREAM));
+				remainingReadLength -= read;
+				this.buffer.pos = read;
+				this.pos += read;
+			}
+		}
+	}
+
+	private void allocate(int size) {
+		this.buffer = new ByteBufferArray(size);
+		this.buffers.add(this.buffer);
+	}
+	
+	private static final class ByteBufferArray {
```

Comment:
I don't see anywhere that you have to copy with a nio ByteBufer that you are able to avoid a copy here?

I understand it is made for NIO originally, but the class is pretty general purpose.

Just my .02, take it for what it is.